mesonbuild/compilers.py

```diff
@@ -300,9 +315,28 @@ def __init__(self, exelist, version):
             self.exelist = exelist
         else:
             raise TypeError('Unknown argument to Compiler')
+        # In case it's been overriden by a child class already
+        if not hasattr(self, 'file_suffixes'):
+            self.file_suffixes = lang_suffixes[self.language]
+        self.default_suffix = self.file_suffixes[0]
+        self.can_compile_suffixes = set(self.file_suffixes)
         self.version = version
         self.base_options = []
 
+    def can_compile(self, src):
+        if hasattr(src, 'fname'):
+            src = src.fname
+        suffix = os.path.splitext(src)[1].lower()
```

Comment:
We can't reliably differentiate between lowercase and uppercase suffixes on all platforms. For instance, on Windows/NTFS depending on what APIs you use, you might get all lowercase, all uppercase, or the actual case that is used by the filesystem. This is because NTFS can store the actual case but APIs don't always.